### ISSUE TITLE
binary cache on macOS is not deterministic 

### ISSUE THREAD

**Issue by autoantwort** [microsoft/vcpkg#19623]
**Describe the bug**
On M1 macs you can build the vcpkg tool as arm64 or x64. But for some libs the resulting lib file has the same architecture as the vcpkg tool regardless of the selected triplet. So if you build the tool as arm64, you get arm64 libs, if you build the tool as x64, you get x64 libs. This problem is getting worse with binary caching, because the hash is the same for the arm64 and the x64 version of the library. So you can't trust the whole binary cache because it maybe contains binary caches for the wrong architecture.

**Environment**
- OS: macOS with M1
- Compiler: apple clang

**Suggested Solution**
Fail when the resulting lib has the wrong architecture (post build check). You can check the architecture of a libs with `lipo -archs ...`.

**Alternative Solution**
Always build the tool as x64. Then you maybe get x64 libs when you use the `arm64-osx` triplet, but then this is deterministic.

**Comment by Be-ing** [comment]
> same architecture as the vcpkg tool

Do you mean that the architecture the vcpkg executable was built for with `./bootstrap-vcpkg.sh`?

**Comment by autoantwort** [comment]
Yes. On arm macs the default is arm64, but you can change it to x64 for example by `arch -x86_64 ./bootstrap-vcpkg.sh`. Then you get an x64 vcpkg tool

**Comment by JackBoosY** [comment]
As more and more users use M1 mac, this matter has become more urgent. 
For bootstrap, we may need to make improvements immediately. 
For pipeline machines, can we use half of the work for arm64-osx?

**Comment by Neumann-A** [comment]
Just like the dumpbin check on windows vcpkg should integrate a lipo check on osx an probably some kind of arch check on normal linux. It is probably way easier to implement on the other platforms since vcpkg can simply assume that the required tools are available on the path.
